cartridges/int_algolia/cartridge/scripts/services/algoliaServiceHelper.js

```diff
@@ -143,6 +141,131 @@ function callJsonService(title, service, params) {
+ * or empty, the key isn't restricted by index name, so only ACL checks matter. Otherwise, we check
+ * if at least one of the "indexes" patterns covers the user-entered indexPrefix.
+ *
+ * @param {dw.svc.Service} service
+ * @param {string} applicationID
+ * @param {string} apiKey
+ * @param {string} indexPrefix         - the user-provided prefix in BM
+ * @param {boolean} isAdminKey
+ * @returns {Object} { error: Boolean, errorMessage: String, warning: String }
+ */
+function validateAPIKey(service, applicationID, apiKey, indexPrefix, isAdminKey) {
+    // 0) Build the required ACL array
+    var requiredACLs;
+    if (isAdminKey) {
+        requiredACLs = ['addObject', 'deleteObject', 'deleteIndex', 'settings'];
+    } else {
+        // minimal read usage is  "search", (initial values for search : "search", "listIndexes", "settings)
+        requiredACLs = ['search'];
+    }
+
+    // 1) Retrieve key info from Algolia
+    var keyResponse = service.call({
+        method: 'GET',
+        url: 'https://' + applicationID + '.algolia.net/1/keys/' + apiKey
+    });
+
+    if (!keyResponse.ok) {
+        return {
+            error: true,
+            errorMessage: Resource.msg('algolia.error.key.validation', 'algolia', null)
+        };
+    }
+
+    var keyData = keyResponse.object.body;
+    var actualACLs = keyData.acl || [];
+
+    // 2) Identify missing vs. excessive ACLs
+    var missingACLs = requiredACLs.slice();
+    var excessiveACLs = [];
+    actualACLs.forEach(function (acl) {
+        var idx = missingACLs.indexOf(acl);
+        if (idx === -1) {
+            excessiveACLs.push(acl);
+        } else {
+            missingACLs.splice(idx, 1);
+        }
+    });
+
+    if (missingACLs.length > 0) {
+        var errMsg = Resource.msgf('algolia.error.missing.permissions', 'algolia', null, missingACLs.join(', '));
+        return {
+            error: true,
+            errorMessage: errMsg,
+            warning: excessiveACLs.length > 0
+                ? Resource.msgf('algolia.warning.excessive.permissions', 'algolia', null, excessiveACLs.join(', '))
+                : ''
+        };
+    }
+
+    // 3) Check index restrictions, if any
+    var restrictedIndexes = keyData.indexes || [];
+    if (restrictedIndexes.length === 0) {
+        // Key is not restricted by indexName => pass
+        return {
+            error: false,
+            errorMessage: '',
+            warning: excessiveACLs.length > 0
+                ? Resource.msgf('algolia.warning.excessive.permissions', 'algolia', null, excessiveACLs.join(', '))
+                : ''
+        };
+    }
+
+
+    var matchedPrefix = restrictedIndexes.some(function (pattern) {
+        // Build a placeholder indexName: we simulate "indexPrefix plus something", e.g. "__products"
+        // so that "test*" would match "test__products" if the prefix = "test".
+        // The simplest is to just add a short suffix to avoid empty string edge case.
```

Comment:
But there is no "empty string edge case"..
If you don't put any index prefix, the cartridge generates one based on the hostname: https://github.com/algolia/algoliasearch-sfcc-b2c/blob/25.1.0/cartridges/int_algolia/cartridge/scripts/algolia/lib/algoliaData.js#L243
This needs to be taken into account too.
Here is a failing scenario based on that:

- Use an API Key with the correct permissions and the following indices: `zzgk_*`
- Try to save the key with "Index Prefix" set to "". Key is not saved :x: . It should be saved, because the indices generated by the cartridge on our sandboxes will be: `zzgk_00x...`, the the key has access to those indices